### PR TITLE
combine write and update to one request

### PR DESCRIPTION
**Pull request by ckiosidis**
Signed-off-by: Babis Kiosidis <ckiosidis@gmail.com>

## Context

We are noticing about 10% errs when talking to gcs, with 404 and 429 (high rate of updates) error codes.

From Google [quotas docs](https://cloud.google.com/storage/quotas#objects)

`There is a write limit to the same object name. This limit is once per second, so writes to the same object name won't scale. For more information, see [Object immutability](https://cloud.google.com/storage/docs/key-terms#immutability).`

Currently it seems that stow will create and update a gcs file when it can be done on a single operation.
The google-go library suggests to [set attributes on the Writer](https://github.com/googleapis/google-cloud-go/blob/5a8c607e3a9a3265887e27cb13f8943f3e3fa23d/storage/storage.go#L1112-L1129) and the Attrs() call should [contain the necessary metadata according to the comment](https://github.com/googleapis/google-cloud-go/blob/0d124aea0fd34510d58167a2b64a17c70c07e43e/storage/writer.go#L344-L348)

What would be the best way to test this change? Any tips?

cc @EngHabu 

_will contribute this back to graymeta/stow if merged_

### CODE REVIEW

File: google/container.go [hunk 1]
```diff
@@ -114,17 +114,25 @@ func (c *Container) Put(name string, r io.Reader, size int64, metadata map[strin
 	}
 
 	w := obj.NewWriter(c.ctx)
+	w.ObjectAttrs.Metadata = merge(w.ObjectAttrs.Metadata, mdPrepped)
 	if _, err := io.Copy(w, r); err != nil {
 		return nil, err
 	}
-	w.Close()
-
-	attr, err := obj.Update(c.ctx, storage.ObjectAttrsToUpdate{Metadata: mdPrepped})
```
Comment (ckiosidis) [inline google/container.go:122]: I wonder if the errors we are seeing on some RPCs are because we are not checking `err = w.Close()`

**Review by RRap0so — APPROVED** [review]

File: google/container.go [hunk 2]
```diff
@@ -114,17 +114,25 @@ func (c *Container) Put(name string, r io.Reader, size int64, metadata map[strin
 	}
 
 	w := obj.NewWriter(c.ctx)
+	w.ObjectAttrs.Metadata = merge(w.ObjectAttrs.Metadata, mdPrepped)
```
Comment (pmahindrakar-oss) [inline google/container.go:117]: The obj.Update does checks arround attrs and for empty string for certain fields eg :
https://github.com/googleapis/google-cloud-go/blob/5a8c607e3a9a3265887e27cb13f8943f3e3fa23d/storage/storage.go#L958

And does workaround by setting `rawObj.NullFields = nullFields`

Will that cause an issue if we now dont use the Update call

Comment (ckiosidis) [reply google/container.go:117]: We only set the metadata in this change.
The rest of the ObjectAttrs we leave as is (which is the same as the previous behaviour)

**Review by EngHabu — APPROVED** [review]
Awesome! Thank you for removing the extra call... I think that makes sense

File: google/container.go [hunk 3]
```diff
@@ -114,17 +114,25 @@ func (c *Container) Put(name string, r io.Reader, size int64, metadata map[strin
...
 		return nil, err
 	}
 
-	return c.convertToStowItem(attr)
+	return c.convertToStowItem(w.Attrs())
+}
+
+func merge(metadata ...map[string]string) map[string]string {
+	res := map[string]string{}
+	for _, mt := range metadata {
```
Comment (lupasarin) [inline google/container.go:130]: If the Metadata is key/value pairs according to [google documentation](https://github.com/googleapis/google-cloud-go/blob/storage/v1.23.0/storage/storage.go#L1361). Why do we need the nested loop? Shouldn't we just:
```
for k, v := range metadata {
    res[k] = v
}
```
?

File: google/container.go (hunk 1 above)
Comment (lupasarin) [reply google/container.go:122]: Great that you added it!

File: google/container.go (hunk 3 above)
Comment (EngHabu) [reply google/container.go:130]: This merges a list of maps into one... so the first loop iterates over the list of maps, the second loop iterates over the keys/values in each of them..

**Pull request merged**